src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -10135,7 +10200,13 @@ private bool ScanExplicitlyTypedLambda(Precedence precedence)
                             }
 
                             continue;
-
+                        case SyntaxKind.ExclamationToken:
+                            SyntaxKind t1k = this.PeekToken(1).Kind;
```

Comment:
this should be taken out of the switch.  it should just be after the `if (this.IsTrueId())`.  If you then see an exclamation you just eat it and hit this switch.

You want the parsing/scanning logic as much as possible to simply be "if i'm here, see if i have this optional `!`.  if so, move past it and have the normal logic from before".